### PR TITLE
feat: fetch single product

### PR DESCRIPTION
**Pull request by Leona-cloud**
## Description
- This PR implements an endpoint that retrieves information about a single product. The response includes product details such as ID, name, description, price, and availability.

## Linked Issue
This PR addresses the following issue [181](https://github.com/hngprojects/hng_boilerplate_nestjs/issues/181)

## Changes Made
API Endpoint Implementation:

Added a GET request endpoint /api/v1/products/{product-id} to fetch a product using the productId provided by the requesting user.
The endpoint validates that the product exists before sending a success or error response to the user.

## Testing
Added unit tests to cover various scenarios:

Successfully fetched products.
Internal Server error if an invalid productId is sent.
Sending a not found error  if product does not exist.

## Documentation

API Documentation:

Endpoint: /api/v1/products/{productId}
Method: GET
Request param:
productId: (string) The unique id of a product.

http
GET /api/v1/products/{product-id}
Content-Type: application/json

Successful Response:
```
{
  "status_code": 200,
  "message": "Product fetched successfully",
  "data": {
    "products": {
      "id": "8b252212-ac35-4e82-b679-6e610f87a120",
      "product_name": "Product 1",
      "description": "Description for Product 1",
      "quantity": 10,
      "price": 100,
      "category": "91a760a6-9b3f-4063-b0a6-a3543b0b6b21",
      "created_at": "2024-07-24T17:38:04.532Z",
      "updated_at": "2024-07-24T17:38:04.532Z"
    }
  }
}
```
Error Response (Product not found):
```
{"error":"Product not found","status_code":404}
```
Error Response (Internal Server Error):
```
{"status_code":500,"message":"Internal server error"}
```

### CODE REVIEW

**Comment by davidolanrewaju** [comment]
Apply openAPI documentation on your routes

File: src/modules/product-category/dto/create-product-category.dto.ts
Comment (Heba-WebDev) [inline src/modules/product-category/dto/create-product-category.dto.ts:1]: Why do you have an empty dto? if it's not used, please remove it.

File: src/modules/product-category/product-category.controller.ts
Comment (Heba-WebDev) [inline src/modules/product-category/product-category.controller.ts:1]: Please remove any endpoint you're not responsible for

File: src/modules/product-category/product-category.service.ts
Comment (Heba-WebDev) [inline src/modules/product-category/product-category.service.ts:1]: Please remove these static services

File: src/modules/product-category/tests/product-category.service.spec.ts
Comment (Heba-WebDev) [inline src/modules/product-category/tests/product-category.service.spec.ts:1]: If you have a service, write few unit tests for it

**Review by Heba-WebDev — CHANGES_REQUESTED** [review]
Please take a look at the comments I left

**Comment by Leona-cloud** [comment]
> Please take a look at the comments I left

I've made an update

**Review by davidolanrewaju — DISMISSED** [review]

**Review by rugue — APPROVED** [review]

File: db/migrations/1721842630604-migration.ts [hunk 1]
```diff
@@ -1,19 +1,28 @@
...
-  name = 'Migration1721761856231';
+export class Migration1721842630604 implements MigrationInterface {
+  name = 'Migration1721842630604';
 
   public async up(queryRunner: QueryRunner): Promise<void> {
-    await queryRunner.query(`CREATE TYPE "public"."user_user_type_enum" AS ENUM('super_admin', 'admin', 'vendor')`);
     await queryRunner.query(
-      `CREATE TABLE "user" ("id" uuid NOT NULL DEFAULT uuid_generate_v4(), "created_at" TIMESTAMP NOT NULL DEFAULT now(), "updated_at" TIMESTAMP NOT NULL DEFAULT now(), "first_name" character varying NOT NULL, "last_name" character varying NOT NULL, "email" character varying NOT NULL, "password" character varying NOT NULL, "is_active" boolean, "attempts_left" integer, "time_left" integer, "user_type" "public"."user_user_type_enum" NOT NULL DEFAULT 'vendor', CONSTRAINT "UQ_e12875dfb3b1d92d7d7c5377e22" UNIQUE ("email"), CONSTRAINT "PK_cace4a159ff9f2512dd42373760" PRIMARY KEY ("id"))`
+      `CREATE TABLE "abstract_base_entity" ("id" uuid NOT NULL DEFAULT uuid_generate_v4(), "created_at" TIMESTAMP NOT NULL DEFAULT now(), "updated_at" TIMESTAMP NOT NULL DEFAULT now(), CONSTRAINT "PK_723c213727b1038d476972ce2f2" PRIMARY KEY ("id"))`
```
Comment (Homoakin619) [inline db/migrations/1721842630604-migration.ts:8]: remove the migration for this table. Its an abstract entity and not meant to be migrated into the database

Comment (Homoakin619) [inline db/migrations/1721842630604-migration.ts:8]: remove the migration for this table. Its an abstract entity and not meant to be migrated into the database

File: db/migrations/1721842630604-migration.ts [hunk 2]
```diff
@@ -23,15 +32,26 @@ export class Migration1721761856231 implements MigrationInterface {
...
     await queryRunner.query(`ALTER TABLE "organisation_preference" DROP CONSTRAINT "FK_2de786f3f89e650581916d67f86"`);
-    await queryRunner.query(`DROP TABLE "organisation"`);
-    await queryRunner.query(`DROP TABLE "organisation_preference"`);
     await queryRunner.query(`DROP TABLE "user"`);
     await queryRunner.query(`DROP TYPE "public"."user_user_type_enum"`);
+    await queryRunner.query(`DROP TABLE "product"`);
+    await queryRunner.query(`DROP TABLE "product_category"`);
+    await queryRunner.query(`DROP TABLE "organisation"`);
+    await queryRunner.query(`DROP TABLE "organisation_preference"`);
+    await queryRunner.query(`DROP TABLE "abstract_base_entity"`);
```
Comment (Homoakin619) [inline db/migrations/1721842630604-migration.ts:55]: remove the drop query for the abstract_base_entity as well

File: src/modules/product-category/product-category.module.ts
Comment (Homoakin619) [inline src/modules/product-category/product-category.module.ts:1]: this is an empty redundant file. remove it

Comment (Leona-cloud) [reply src/modules/product-category/product-category.module.ts:1]: I've made changes

**Review by Heba-WebDev — APPROVED** [review]

File: src/modules/products/products.service.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+import { Repository } from 'typeorm';
+
+@Injectable()
+export class ProductsService {
+  private readonly logger = new Logger(ProductsService.name);
+
+  constructor(@InjectRepository(Product) private productRepository: Repository<Product>) {}
+
+  async fetchSingleProduct(productId: string) {
+    this.logger.log(`Attempting to retrieve product with id: ${productId}`);
```
Comment (Homoakin619) [inline src/modules/products/products.service.ts:13]: remove this

File: src/modules/products/products.service.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+    const productExists = await this.productRepository.findOne({
+      where: {
+        id: productId,
+      },
+      relations: ['category'],
+    });
+    if (!productExists) {
+      this.logger.log(`Product with id: ${productId} does not exist`);
```
Comment (Homoakin619) [inline src/modules/products/products.service.ts:22]: remove this as well

File: src/modules/products/entities/product.entity.ts
Comment (Heba-WebDev) [inline src/modules/products/entities/product.entity.ts:1]: Please specify the nullability of columns, nullable: true/false

Comment (Leona-cloud) [reply src/modules/products/entities/product.entity.ts:1]: Okay, done.

**Review by Heba-WebDev — APPROVED** [review]
Looks good to me

**Review by davidolanrewaju — APPROVED** [review]

**Review by Cobidex — APPROVED** [review]

**Pull request merged**